### DISCUSSION TITLE
Wait for array.map iterations in Promise.all [duplicate]

### DISCUSSION THREAD

**Discussion by strong-728** [community/community#36991] [Other Feature Feedback, Questions, & Ideas] [answered]
have the following code that should add items for customers if they dont already exist. Execution should be in parallel.
 await Promise.all(
    customers.map(async (customer) => {
        return customer.items.map(async (item) => {
            return new Promise(async (resolve) => {
                const productExists = someArray.some(
                    (arrayValue) => arrayValue === item.id
                );
                if (!productExists) {
                    logger.info(
                    `customer item ${item.id} does not exist, creating...`
                    );
                    await createCustomerItem(item.id);
                    logger.info(`customer item ${item.id} created.`);

                    someArray.push(item.id);
                } else {
                    logger.info(`customer item ${item.id} already exists, skipping...`);
                }
                resolve(true);
            });
        });
    })

);

logger.info(`All items should now be present`);

**Comment by codeshare007** [comment, marked as answer]
Try with [flatMap](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Array/flatMap):

```
await Promise.all(
    customers.flatMap(async (customer) => {
        return customer.items.map(async (item) => {
```
Instead of returning an array of array of Promises, it will flatten the content to a simple array of Promises, which is what Promise.all expects.

N.B. It is not apparent from your question if the grouping of items per customer needs to be conserved. If it does, note that this solution changes the data structure to a flattened list, so you lose the grouping. Either add some customerId in your items, or try @blex's suggestion in comments.
